### PR TITLE
Artillery placement

### PR DESCRIPTION
**Pull request by martkartasev**
Full list of changes too long to specify.
Adds Team and formation level AI control capability. 
Adds AI team artillery control using that capability.
Changes the related casting behavior to be dependent on the team level tactic, so its all interconnected.

### CODE REVIEW

File: CSharpSourceCode/BattleMechanics/AI/AgentBehavior/AgentCastingBehavior/ArtilleryPlacementCastingBehavior.cs [hunk 1]
```diff
@@ -1,15 +1,71 @@
...
+
+            var castingPosition = ((AdjacentAoETacticalBehavior) TacticalBehavior)?.CastingPosition;
+            if (castingPosition.HasValue && Agent.Position.AsVec2.Distance(castingPosition.Value.AsVec2) > 10) return;
+
+            base.Execute();
+        }
+
+        protected override Target UpdateTarget(Target target)
+        {
+            var width = Math.Max(target.TacticalPosition.Width, 35);
```
Comment (Z3rca) [inline CSharpSourceCode/BattleMechanics/AI/AgentBehavior/AgentCastingBehavior/ArtilleryPlacementCastingBehavior.cs:37]: Are we sure this is never =0?

**Review by hunharibo — APPROVED** [review]
Looks fine to me.

File: CSharpSourceCode/Extensions/MissionExtensions.cs [hunk 2]
```diff
@@ -13,6 +15,21 @@ public static bool IsPlayerInSpellCasterMode(this Mission mission)
...
         
+        public static void AddMissionLogicAtIndexOf(this Mission mission, MissionLogic missionCombatantsLogic, MissionLogic torMissionCombatantsLogic)
+        {
+            var behaviorIndex = mission.MissionBehaviors.FindIndex(item => item.GetType() == missionCombatantsLogic.GetType());
+            var logics = Traverse.Create(mission).Field("_missionLogics").GetValue() as List<MissionLogic>;
+            var logicsIndex = logics.FindIndex(item => item.GetType() == missionCombatantsLogic.GetType());
+            mission.RemoveMissionBehavior(missionCombatantsLogic);
+
+            mission.AddMissionBehavior(torMissionCombatantsLogic); //TODO: Need to call this so that .mission is set on the behavior
+            mission.MissionBehaviors.Remove(torMissionCombatantsLogic); //TODO: Then we remove without calling the mission.RemoveMissionBehavior, as it sets Mission to null.
```
Comment (Z3rca) [inline CSharpSourceCode/Extensions/MissionExtensions.cs:26]: I don't get this still *^^

File: CSharpSourceCode/BattleMechanics/AI/AgentBehavior/AgentCastingBehavior/ArtilleryPlacementCastingBehavior.cs (hunk 1 above)
Comment (Z3rca) [reply CSharpSourceCode/BattleMechanics/AI/AgentBehavior/AgentCastingBehavior/ArtilleryPlacementCastingBehavior.cs:37]: sorry read min

**Pull request merged**